### ISSUE TITLE
[Feature] Update bundled fim to latest version (and unbundle?)

### ISSUE THREAD

**Issue by sanjayankur31** [NeuralEnsemble/elephant#471]
**Feature**

The version of fim bundled in the sources is 0.4.x, but the latest version of fim is now at 6.x. Would it be possible to please:

- use the version of fim on pypi (which removes the need to bundle it): https://pypi.org/project/pyfim/
- or if bundling is necessary, could we please update the bundled copy to keep it in sync with upstream fim releases (otherwise we're going to be using an older obsolete version of fim)

**Motivation**

Since fim is available on pypi, it's probably best to use this and not bundle a copy. Bundling creates all sorts of issues, one of which is that the bundled copy needs to be manually kept in sync with upstream releases. So if the pypi version can be used, we just try and use that.

**Alternatives**

**Additional context**

We try to use system libraries as much as possible in Fedora, so we've got pyfim packaged separately and ideally we'd like to use that instead of the bundled copy here. At the moment, though, it looks like some signatures have changed in fim in b/w version 4.x (bundled here) and version 6.x (upstream latest, and packaged in Fedora), which is breaking some tests:

Here's an example:
```
>               fpgrowth_output = fim.fpgrowth(                                                                                         
                    tracts=transactions,                                                                                                
                    target=target,                                                                                                      
                    supp=-min_c,                                                                                                        
                    zmin=min_z,                                                                                                         
                    zmax=max_z,                                     
                    report='a',                                                                                                         
                    algo='s',                                       
                    winlen=winlen,                                                                                                      
                    threads=0,                                                                                                          
                    verbose=4)                                                                                                          
E               TypeError: 'winlen' is an invalid keyword argument for this function             
```

The new signature indeed does not include this:

```
Help on built-in function fpgrowth in module fim:

fpgrowth(...)
    fpgrowth (tracts, target='s', supp=10, zmin=1, zmax=None, report='a',
              eval='x', agg='x', thresh=10, prune=Nobe, algo='s', mode='',
              border=None)

```

**Comment by mdenker** [comment]
Hi @sanjayankur31, thanks for your input. Indeed, having `pyfim` available on pypi now is extremely useful. However, in this case, the situation is a little different.

In prior versions (<0.11.0), Elephant indeed used `pyfim` and relied on having the binary`fim.so` available (e.g., from the author's web page). However, in the current version, we have had contributions to optimize the `fpgrowth` algorithm specifically using aspects of  the `spade` analysis that requires this function. For historical reasons, the code retains strong similarity to the call signature of `pyfim` on the python side, but on the backend the original C pyfim implementation was replaced  by a new C++ implementation of fpgrowth implementing these specific optimizations. The optimizations significantly speed up the computation. However, due to the specific nature of this implementation of fpgrowth (which is *much* less generic and feature-rich compared to pyfim), we are not sure it can be packaged stand-alone, but it is currently rather tightly linked to `spade` analysis.

However, I do take from your suggestion that we could include pyfim as a dependency to implement a fall-back for systems where compilation of the C++ code is not possible, but `pyfim` exists. This will require a bit of time, however, maybe this can be done for the 0.12.0 release.

**Comment by mdenker** [comment]
Hi @sanjayankur31, if the C++ extension poses problems, please let us know. In the worst case, it's possible to skip building of the module, which will however result in a much slower Python-only fall-back. We'll check if we can speed this fallback up for the upcoming release by using the "regular" `pyfim` package.

I will leave this issue open until this is investigated and to report back the outcome.

**Comment by sanjayankur31** [comment]
The C++ extension builds fine, so I'm using that for the Fedora build now. The compilation would error out using the default compilation flags used in `setup.py` here on GCC 12, but with the default Fedora build flags, it builds fine. So, I'm just carrying this tiny downstream patch:

https://github.com/sanjayankur31/elephant/commit/9e38be2effa193c32a5cf44601f51612c15137f1

Here is the compilation error bit:

```
  gcc -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unw
ind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -W
error=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwi
nd-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -We
rror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwin
d-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-
gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-har
dened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstac
k-clash-protection -fcf-protection -fPIC -Ielephant/spade_src/include -I/usr/include/python3.10 -c elephant/spade_src/src/fim.cpp -o bui
ld/temp.linux-x86_64-3.10/elephant/spade_src/src/fim.o -DMODULE_NAME=fim -DUSE_OPENMP -DWITH_SIG_TERM -Dfim_EXPORTS -O3 -pedantic -Wextr
a -Weffc++ -Wunused-result -Werror -fopenmp -std=gnu++17
  In file included from elephant/spade_src/src/fim.cpp:41:
  elephant/spade_src/include/FPGrowth.h: In constructor ‘FPGrowth::FPGrowth(Transactions&, Support, uint32_t, uint32_t, ItemC, uint32_t,
 uint32_t, int32_t)’:                                                                                                                     elephant/spade_src/include/FPGrowth.h:198:37: error: loop variable ‘p’ of type ‘const RefPair&’ {aka ‘const std::pair<unsigned int, st
d::shared_ptr<FrequencyRef> >&’} binds to a temporary constructed from type ‘std::pair<const unsigned int, std::shared_ptr<FrequencyRef> >’ [-Werror=range-loop-construct]
    198 |                 for (const RefPair& p : F)    
        |                                     ^
  elephant/spade_src/include/FPGrowth.h:198:37: note: use non-reference type ‘const RefPair’ {aka ‘const std::pair<unsigned int, std::sh
ared_ptr<FrequencyRef> >’} to make the copy explicit or ‘const std::pair<const unsigned int, std::shared_ptr<FrequencyRef> >&’ to preven
t copying                                                                                                                                 cc1plus: all warnings being treated as errors                                                                                           error: command '/usr/bin/gcc' failed with exit code 1                                                                                   error: subprocess-exited-with-error                                                                                                                                                                                                                                             × Building wheel for elephant (pyproject.toml) did not run successfully.                                                                │ exit code: 1                                                    
  ╰─> See above for output.
                                                                    
```

**Comment by Moritz-Alexander-Kern** [comment]
This was addressed in PR #508.

See:
https://github.com/NeuralEnsemble/elephant/blob/df8e96212e121d5931bab6c9ae82e059c91ec2ad/elephant/spade_src/include/FPGrowth.h#L198

Fix will be included in v0.11.2 of elephant.
